### PR TITLE
Fix race in TestStartMachineWithManualMachine (concurrent map access)

### PR DESCRIPTION
**Pull request by benhoyt**
I believe the production code itself is not racey, as everything goes through the big select in Firewaller.loop(). However, this test bypasses that and calls the unexported startMachine and getMachineds methods, and the background goroutine running loop() might be writing the machineds map at the same time.

To fix, remove firewaller.StartMachine() and firewaller.GetMachineds() altogether, and add a "notify function" for when the Firewaller actually starts watching a machine (shouldn't happen for manual machines).

Also, two drive-bys:

1) No need for return after c.Fatal() calls
2) In State.StartSync(), syncable.StartSync() is now a no-op as the txnLogWatcher (a HubWatcher or TxnWatcher) doesn't have a StartSync method anymore.

Here's an example of a race failure (from [here](https://jenkins.juju.canonical.com/job/Unit-RunUnitTests-race-amd64/77/testReport/junit/github/com_juju_juju_worker_firewaller/TestAll/)):

```
WARNING: DATA RACE
Write at 0x00c0004ddc50 by goroutine 134:
  runtime.mapassign_faststr()
      /usr/local/go/src/runtime/map_faststr.go:202 +0x0
  github.com/juju/juju/worker/firewaller.(*Firewaller).startMachine()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/worker/firewaller/firewaller.go:460 +0x625
  github.com/juju/juju/worker/firewaller.StartMachine()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/worker/firewaller/export_test.go:9 +0x82a
  github.com/juju/juju/worker/firewaller_test.(*InstanceModeSuite).TestStartMachineWithManualMachine()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/worker/firewaller/firewaller_test.go:543 +0x805
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:321 +0xd3
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/vendor/gopkg.in/check.v1/check.go:781 +0xa0a
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/vendor/gopkg.in/check.v1/check.go:675 +0xd9

Previous read at 0x00c0004ddc50 by goroutine 173:
  runtime.mapaccess2_faststr()
      /usr/local/go/src/runtime/map_faststr.go:107 +0x0
  github.com/juju/juju/worker/firewaller.(*Firewaller).machineLifeChanged()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/worker/firewaller/firewaller.go:1090 +0x141
  github.com/juju/juju/worker/firewaller.(*Firewaller).loop()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/worker/firewaller/firewaller.go:286 +0x656
  github.com/juju/juju/worker/firewaller.(*Firewaller).loop-fm()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/worker/firewaller/firewaller.go:271 +0x41
  github.com/juju/worker/v2/catacomb.runSafely()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/vendor/github.com/juju/worker/v2/catacomb/catacomb.go:295 +0x70
  github.com/juju/worker/v2/catacomb.Invoke.func3()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/vendor/github.com/juju/worker/v2/catacomb/catacomb.go:116 +0xa4
  gopkg.in/tomb%2ev2.(*Tomb).run()
      /home/jenkins/workspace/Unit-RunUnitTests-race-amd64/_build/src/github.com/juju/juju/vendor/gopkg.in/tomb.v2/tomb.go:163 +0x4c
```

### CODE REVIEW

File: worker/firewaller/firewaller.go [hunk 1]
```diff
@@ -94,6 +94,10 @@ type Config struct {
 	Logger Logger
 
 	CredentialAPI common.CredentialAPI
+
+	// WatchMachineNotify is called when the Firewaller starts watching the
```
Comment (wallyworld) [inline worker/firewaller/firewaller.go:98]: Please include a phrase that this is only used for tests

File: worker/firewaller/firewaller.go [hunk 2]
```diff
@@ -157,6 +161,8 @@ type Firewaller struct {
 	logger                     Logger
 
 	cloudCallContext context.ProviderCallContext
+
+	watchMachineNotify func(tag names.MachineTag)
```
Comment (wallyworld) [inline worker/firewaller/firewaller.go:165]: Maybe the "only for tests" comment goes here (see above).

**Review by wallyworld — APPROVED** [review]
Thanks.
Maybe consider a type definition for func(tag names.MachineTag)

**Comment by benhoyt** [comment]
!!build!!

**Comment by benhoyt** [comment]
!!build!!

**Comment by benhoyt** [comment]
$$merge$$

**Pull request merged**